### PR TITLE
Setup fake tensor and symbolic shapes once at beginning of AOTAutograd

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85260
* #85264
* __->__ #85233
* #85301

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D39662822](https://our.internmc.facebook.com/intern/diff/D39662822)

### CODE REVIEW

File: functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -394,34 +400,31 @@ def create_aot_dispatcher_function(
...
-                    # instead of memoizing the conversion. this has the same
-                    # problem of models that resize their inputs described below,
-                    # but fixes an issue with tied parameters.
-                    # TODO: more full fix
-                    if id(x) in seen_args:
-                        with torch.utils._mode_utils.no_dispatch():
-                            x = x.detach().requires_grad_(x.requires_grad)
-                    seen_args.add(id(x))
-                    return mode.from_tensor(x)
+                    return fake_mode.from_tensor(x, shape_env=shape_env)
```
Comment (eellison) [inline functorch/_src/aot_autograd.py:427]: I couldnt figure out a good small representative test for this but without this fix like half of the bert models break, I dont think we can just get rid of this logic atm

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -203,9 +203,40 @@ Tensor& set_storage_cpu_(Tensor& result, Storage storage, int64_t storage_offset
   return result;
 }
 
-Tensor& set_(Tensor& result, const Tensor& storage, int64_t storage_offset, IntArrayRef size, IntArrayRef stride) {
+Tensor& set_storage_meta__symint(Tensor& result, Storage storage, c10::SymInt storage_offset, c10::SymIntArrayRef size, c10::SymIntArrayRef stride) {
```
Comment (voznesenskym) [inline aten/src/ATen/native/TensorShape.cpp:206]: can these refs be const?

**Review by wconstab — COMMENTED** [review]
for most part this LGTM except I defer to @eellison's comment about the issue with bert.
Also it wouldn't hurt to clean up a couple of TODOs in set_storage_meta

File: aten/src/ATen/native/TensorShape.cpp [hunk 3]
```diff
@@ -203,9 +203,40 @@ Tensor& set_storage_cpu_(Tensor& result, Storage storage, int64_t storage_offset
...
+      contiguous_strides.at(last_idx) = 1;
+      for (auto i = last_idx - 1; i >= 0; --i) {
+        // TODO: max with 1
+        contiguous_strides.at(i) = contiguous_strides.at(i+1) * size.at(i+1);
+      }
+    }
+    stride = contiguous_strides;
+  }
+
+  // TODO: implement this
```
Comment (wconstab) [inline aten/src/ATen/native/TensorShape.cpp:225]: how big of a deal is this? worth getting it done now or going to defer?

the TODOs above seem more like things you could knock out now

File: functorch/_src/aot_autograd.py (hunk 1 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:427]: @eellison that doesn't make sense, I have tested BERT_pytorch with this change and it passes... which models?

File: aten/src/ATen/native/TensorShape.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/native/TensorShape.cpp:206]: no, they cannot. This is a long standing missing staircase that cannot be fixed without doing a very large and annoying c++ codemod

File: aten/src/ATen/native/TensorShape.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/native/TensorShape.cpp:225]: we can do it now

File: functorch/_src/aot_autograd.py (hunk 1 above)
Comment (eellison) [reply functorch/_src/aot_autograd.py:427]: `python benchmarks/huggingface.py --ci -d cuda --float32 --backend=aot_nop --training --only=RobertaForCausalLM`

Comment (eellison) [reply functorch/_src/aot_autograd.py:427]: https://github.com/pytorch/pytorch/issues/85093

Comment (ezyang) [reply functorch/_src/aot_autograd.py:427]: works on the symbolic-shapes branch (modulo accuracy problems lol)

```
torchdynamo.utils: [WARNING] RMSE (res-fp64): 0.00034, (ref-fp64): 0.00016
INCORRECT
Command exited with non-zero status 255
17.88user 18.49system 0:19.60elapsed 185%CPU (0avgtext+0avgdata 2531928maxresident)k
0inputs+296outputs (0major+645318minor)pagefaults 0swaps
[ezyang@devgpu001.pnb3 ~/local/pytorch-tmp (symbolic-shapes)]$ (cd ../torchdynamo && pp TORCH_SHOW_CPP_STACKTRACES=1 AOT_FX_GRAPHS_JOINT=1 TORCHDYNAMO_DYNAMIC_SHAPES=0 AOT_DYNAMIC_SHAPES=0 time python benchmarks/huggingface.py --only RobertaForCausalLM --accuracy-aot-nop --training) 
```

**Comment by ezyang** [comment]
roberta accuracy error is fixed in parent commit on stack now

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39662822).

**Review by wconstab — APPROVED** [review]
